studio/data/table-rows/table-rows-query.ts

```diff
@@ -0,0 +1,116 @@
+import { QueryKey, UseQueryOptions } from '@tanstack/react-query'
+import { Filter, Query, Sort, SupaRow, SupaTable } from 'components/grid'
+import { ExecuteSqlData, useExecuteSqlPrefetch, useExecuteSqlQuery } from '../sql/execute-sql-query'
+import { getPagination } from '../utils/pagination'
+import { formatFilterValue } from './utils'
+
+type GetTableRowsArgs = {
+  table?: SupaTable
+  filters?: Filter[]
+  sorts?: Sort[]
+  limit?: number
+  page?: number
+}
+
+export const getTableRowsSqlQuery = ({
+  table,
+  filters = [],
+  sorts = [],
+  page,
+  limit,
+}: GetTableRowsArgs) => {
+  const query = new Query()
+
+  if (!table) {
+    return ``
+  }
+
+  const enumArrayColumns = table.columns
+    .filter((column) => {
+      return (column?.enum ?? []).length > 0 && column.dataType.toLowerCase() === 'array'
+    })
+    .map((column) => column.name)
+
+  let queryChains =
+    enumArrayColumns.length > 0
+      ? query
+          .from(table.name, table.schema ?? undefined)
+          .select(`*,${enumArrayColumns.map((x) => `"${x}"::text[]`).join(',')}`)
+      : query.from(table.name, table.schema ?? undefined).select()
+
+  filters
+    .filter((x) => x.value && x.value != '')
+    .forEach((x) => {
+      const value = formatFilterValue(table, x)
+      queryChains = queryChains.filter(x.column, x.operator, value)
+    })
+  sorts.forEach((x) => {
+    queryChains = queryChains.order(x.column, x.ascending, x.nullsFirst)
+  })
+
+  // getPagination is expecting to start from 0
+  const { from, to } = getPagination((page ?? 1) - 1, limit)
+  const sql = queryChains.range(from, to).toSql()
+
+  return sql
+}
+
+export type TableRows = {
+  rows: SupaRow[]
+}
+
+export type TableRowsVariables = GetTableRowsArgs & {
+  projectRef?: string
+  connectionString?: string
+  queryKey?: QueryKey
+}
+
+export type TableRowsData = TableRows
+export type TableRowsError = unknown
+
+export const useTableRowsQuery = <TData extends TableRowsData = TableRowsData>(
+  { projectRef, connectionString, queryKey, table, ...args }: TableRowsVariables,
+  options: UseQueryOptions<ExecuteSqlData, TableRowsError, TData> = {}
+) =>
+  useExecuteSqlQuery(
+    {
+      projectRef,
+      connectionString,
+      sql: getTableRowsSqlQuery({ table, ...args }),
+      queryKey: [
+        ...(queryKey ?? []),
+        { table: { name: table?.name, schema: table?.schema }, ...args },
+      ],
+    },
+    {
+      select(data) {
+        const rows = data.result.map((x: any, index: number) => {
+          return { idx: index, ...x } as SupaRow
+        })
+
+        return {
+          rows,
+        } as TData
+      },
+      enabled: typeof projectRef !== 'undefined' && typeof table !== 'undefined',
+      ...options,
+    }
+  )
+
+export const useTableRowsPrefetch = ({
```

Comment:
realised the prefetch methods aren't being used - do we need this?
or more so, when do you reckon we'd use this over the useTableRowsQuery

Reply:
All the prefetch methods are there for future use. We'll be able to put the `prefetch` function they return on hover to the page links, so the data starts loading while the page js is loading.

Because they're not currently used, just as above, I'm happy to remove these, for now, to clean things up and then add them back when we come to using them. wdyt?

Reply:
no worries! but could you leave a comment about this too? so that we'll know what they'll be used for